### PR TITLE
8702 purchase extractor

### PR DESCRIPTION
**Pull request by DevneyHamilton**

### CODE REVIEW

File: app/models/accounting/quickbooks/transaction_extractor.rb [hunk 1]
```diff
@@ -17,86 +43,63 @@ def extract!
...
           # skip if line item does not have an account in Madeline
           next unless acct
 
           txn.line_item_with_id(li['id'].to_i).assign_attributes(
             account: acct,
             amount: li['amount'],
-            posting_type: li['journal_entry_line_detail']['posting_type']
+            posting_type: li[qb_li_detail_key]['posting_type']
           )
         end
```
Comment (DevneyHamilton) [inline app/models/accounting/quickbooks/transaction_extractor.rb]: Is now a good time to combine this logic (for DRY) or should we wait and see how the other subclasses shake out?

File: spec/models/accounting/quickbooks/data_extractor_spec.rb [hunk 2]
```diff
@@ -7,16 +7,24 @@
...
 
+    it "calls the Purchase extractor class for purchase" do
+      txn = create(:accounting_transaction, qb_object_type: 'Purchase')
+      extractor = double("extractor")
+      expect(Accounting::Quickbooks::PurchaseExtractor).to receive(:new).with(txn) { extractor }
+      expect(extractor).to receive(:extract!)
+      described_class.new(txn).extract!
+    end
+
     # Eventually account extraction should move to another subclass
```
[the remark this answers is not in the document]
Comment (meltheadorable) [reply spec/models/accounting/quickbooks/data_extractor_spec.rb]: Maybe? Yes. We don't extract much data from accounts right now, so it's all done by hand in `create_or_update_from_qb_object` on the model but it's possible that becomes relevant again when we do the bigger refactor which changes all the fetchers. It's certainly not worth keeping commented out code for a theoretical model sitting around though, and it's not a particularly hard test to write, so feel free to delete this.

File: spec/models/accounting/quickbooks/purchase_extractor_spec.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
+require 'rails_helper'
+
+describe Accounting::Quickbooks::PurchaseExtractor, type: :model do
```
[the remark this answers is not in the document]
Comment (meltheadorable) [reply spec/models/accounting/quickbooks/purchase_extractor_spec.rb]: Probably fine, unless you noticed a difference between it and the other examples that you think is super relevant and we should also test.

[the remark this answers is not in the document]
Comment (meltheadorable) [reply spec/models/accounting/quickbooks/purchase_extractor_spec.rb]: Anonymize it more though, we don't want actual vendors, names or orgs that work with TWW to be in here, even if the transaction amounts and dates and such are fictionalized.

File: app/models/accounting/quickbooks/transaction_extractor.rb [hunk 4]
```diff
@@ -75,7 +78,7 @@ def calculate_amount
         # However, we define our 'amount' as the sum of the change_in_interest and change_in_principal,
         # which are computed from a special subset of line items (see the Transaction model for more detail).
         # This may mean that our amount may differ from the amount shown in Quickbooks for this transaction,
-        # but that is ok.
+        # but that is ok. QUESTION: is this still ok?
```
Comment (meltheadorable) [inline app/models/accounting/quickbooks/transaction_extractor.rb]: I think this is still okay, but we can discuss.

**Pull request merged**